### PR TITLE
[Kernel] Add losslessly packed BF16 lm_head backend

### PR DESCRIPTION
**Pull request by adenzhou1350**
## Purpose

This PR adds an opt-in, losslessly packed BF16 backend for unquantized language
model heads during single-token decode.

The backend stores an exact auxiliary encoding of the BF16 weights and uses a
Triton projection kernel for M=1. `ParallelLMHead` installs it as a decorator
around the existing unquantized method, so unsupported dtypes, shapes, bias,
platforms, and larger batches retain the existing path. The default remains
`torch`.

The packing pass is chunked to bound temporary memory, checks the materialized
storage ratio and available device memory, preserves outlier blocks verbatim,
and participates in vLLM's JIT warmup.

### Duplicate-work check

This does not duplicate #52355, which adds OOT pluggable-layer coverage for
embedding/lm-head/logits-processor classes but does not implement a BF16 weight
format or projection kernel. It also does not duplicate #48870, which targets
tied, quantized WNA16 lm-head weights; this change targets unquantized BF16
heads and uses an exact encoding.

## Test Plan

```bash
.venv/bin/python -m pytest \
  tests/kernels/core/test_packed_bf16_lm_head.py -v
pre-commit run --files <all six changed files>
# all applicable hooks passed
```

The model-level comparison uses Qwen3.5-0.8B with a GPTQ-Marlin backbone and an
unquantized BF16 lm-head, TP=1, max_num_seqs=1, six natural prompts, one warmup
plus three measured runs per prompt, and 64 generated tokens per run.

## Test Result

The CUDA test suite passed: `7 passed`.

The full changed-files pre-commit run passed, including ruff-format,
ruff-check, typos, Python 3.10 mypy, SPDX, lazy-import, filename,
forbidden-import, torch-CUDA-call, configuration validation, and suggestion
checks. Hooks unrelated to the changed file types were skipped as expected.

### Kernel benchmark

RTX 4060 Laptop (SM89), BF16, shape
`(1, 1024) x (248320, 1024)^T`:

```text
packed storage: 0.772 of dense
torch: 2.153 ms
packed: 1.683 ms
speedup: 1.279x
```

Command:

```bash
.venv/bin/python benchmarks/kernels/benchmark_packed_bf16_lm_head.py \
  --vocab-size 248320 --hidden-size 1024
```

### Model evaluation

| metric | stock | packed | change |
|---|---:|---:|---:|
| Mean median E2E | 297.356 ms | 272.708 ms | 1.090x / -8.29% |
| Mean median TPOT | 4.0672 ms | 3.7256 ms | 1.092x / -8.40% |

All six stock/candidate greedy token sequences matched exactly. The test used
the V1 model runner on WSL because the current V2 runner requires UVA, which
was unavailable in this environment.

The patch is currently rebased onto `f4eccda`, where the CUDA tests were rerun.
The reported performance run was recorded on `52bc900`; the two later upstream
commits touch only the HY V4 kernel and multimodal renderer, not this patch's
paths. A CUDA 13 nightly wheel for `f4eccda` was not available yet, so the
editable current Python source used the precompiled extension from `28e605f`.
The intervening binary-relevant change only adds SM110 to a CMake architecture
list and does not touch this SM89, non-MTP execution path.

The real model's auxiliary layout was 375.519 MiB (0.774 of dense), reducing
available KV-cache memory by approximately 0.39 GiB. This memory/latency
tradeoff and the current SM89-only performance evidence are why the backend is
opt-in.

## AI assistance disclosure

AI assistance was used to inspect the existing architecture, implement the
initial patch, and prepare tests and benchmark analysis.

The human submitter reviewed every changed line, confirmed the encoding and
fallback behavior, and reran the reported tests before marking the PR ready for
review.

## Checklist

- [x] The purpose and compatibility boundary are documented.
- [x] The test plan and measured results are included.
- [x] Correctness and memory/performance tradeoffs are reported.
- [x] Human line-by-line review and local rerun completed before ready-for-review.

### CODE REVIEW

**Comment by adenzhou1350** [comment]
Latest-main-at-test requalification update (parent `719284fe`, review head `b72f0008`):

- RTX 4060 Laptop / SM89, BF16, shape `(1, 248320, 1024)`.
- Two fresh-process kernel repeats were correctness-PASS and measured 1.442x / 1.411x versus torch.
- Packed CUDA/contract suite: 9 passed; fallback contract tests: 2 passed.
- Broader weight-tying/head-dtype run: 13 passed; one V2-engine test was environment-blocked because UVA is unavailable in this WSL setup, with no candidate-path failure.
- Both commits carry the submitter DCO sign-off; changed-file lint/format/type/policy hooks passed.

Claim boundary: this remains an opt-in SM89 large-vocabulary, single-token decode optimization. The historical RTX 4060 model-level result was 1.09x with exact greedy token identity, but the separate RTX 5090 whole-model portfolio did not clear the materiality gate (pooled 1.0058x; conservative 0.9871x), so I am not claiming general RTX 5090 or general vLLM acceleration.

@22quinn, when you have time, a review of the packed representation / fallback boundary would be especially helpful. The branch is currently out of date again as main continues moving; I will refresh and rerun the source-sensitive checks before merge rather than treating this comment as final merge qualification.

**Comment by adenzhou1350** [comment]
Rebased this PR onto upstream `ca67438c08`; the new head is `364ea0435f`.

The only conflict was the shared insertion point in `KernelConfig`. Both upstream's per-quant linear-backend overrides and this PR's opt-in lm-head configuration are preserved. The packed kernel and fallback implementation are unchanged; a regression assertion now checks that both configurations coexist.

Validation: Ruff check/format passed for all 10 changed Python files, `git diff --check` passed, and real `KernelConfig` normalization/coexistence/default assertions passed with CUDA hidden. The selected packed-head pytest module was blocked during collection because the fresh local worktree has no FlashAttention native extensions; no tests from that run executed. No extension build, GPU run, or new performance measurement was performed.

The earlier SM89 measurements remain evidence at their originally recorded revisions, not a claim of full current-main native requalification. The opt-in scope and the disclosed RTX5090 whole-model non-result are unchanged.

**Comment by adenzhou1350** [comment]
Follow-up validation on `364ea0435f59`: **all 23 cases across the three affected test files passed** on RTX 5090 — 22 kernel/weight-tying/head-dtype cases, then the registered OPT-125M FP32-head E2E test separately. The latter loaded real weights with the V2 model runner/FA2 and produced finite logprobs (no NaNs).

```bash
.venv/bin/python -m pytest -v tests/kernels/core/test_packed_bf16_lm_head.py tests/model_executor/model_loader/test_weight_tying.py tests/v1/sample/test_head_dtype.py -k 'not test_fp32_head_e2e_no_nan'
.venv/bin/python -m pytest -v tests/v1/sample/test_head_dtype.py::test_fp32_head_e2e_no_nan
```

Source/tests were unchanged, with imported paths and source hashes checked. Environment: Python 3.12, Torch 2.13, Triton 3.7.1, FlashInfer 0.6.18.post1. This used exact PR Python/Triton source with reused `g7ee8a6dd0` native binaries, **not an exact-head native build or the complete official CI environment**. OPT-125M came from a hash-checked ModelScope mirror at revision `3a786641e96e9e4d24ab6e2155040a6cfb08e57b`, exposed locally for the unchanged test; HF byte equivalence was not independently verified.

AI-assisted correctness follow-up only; no new performance claim. GPU cleanup was verified.
